### ISSUE TITLE
Supplementary_Table_S3.py

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/loh-work-disability-experimental-notes#96]
"""Supplementary Table S3. Nonlinear-indicator evaluability and boundaries.

This page summarizes Figure 2 indicator evaluability, boundary operational t*,
cluster-specific distributions, and the raw-to-display K-means label mapping.

Prerequisites: run 00_Common_Setup.py, Figure 1, and Figure 2.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
from scipy.ndimage import gaussian_filter1d
from sklearn.cluster import KMeans
from sklearn.impute import SimpleImputer

required_globals = (
    "ANALYSIS_READY_FILE", "OUTPUT_DIR", "ANALYSIS_READY_N",
    "TRAJECTORY_COHORT_N", "EXPECTED_CLUSTER_COUNTS",
    "EXPECTED_ANALYSIS_SHA256", "REFERENCE_SIGMA", "RANDOM_STATE",
    "KMEANS_N_INIT", "read_csv_robust", "verify_rows", "verify_sha256",
    "add_analysis_row_id", "psa_columns_in_time_order",
)
missing_globals = [name for name in required_globals if name not in globals()]
if missing_globals:
    raise RuntimeError("Run 00_Common_Setup.py first. Missing: " + ", ".join(missing_globals))

CLUSTER_FILE = OUTPUT_DIR / "Figure1" / "Figure1_Cluster_Assignments_1434.csv"
INDICATOR_FILE = OUTPUT_DIR / "Figure2" / "Figure2_Per_Patient_Nonlinear_Indicators.csv"
for path in (CLUSTER_FILE, INDICATOR_FILE):
    if not path.exists():
        raise FileNotFoundError(f"Missing prerequisite output: {path}")

OUTDIR = OUTPUT_DIR / "Supplementary_Table_S3"
OUTDIR.mkdir(parents=True, exist_ok=True)

clusters = pd.read_csv(CLUSTER_FILE)[["Analysis_Row_ID", "Cluster"]]
indicators = pd.read_csv(INDICATOR_FILE)
if indicators["Cluster"].value_counts().sort_index().to_dict() != EXPECTED_CLUSTER_COUNTS:
    raise AssertionError("Canonical cluster counts do not match the manuscript")

def median_iqr(series: pd.Series, decimals: int) -> str:
    values = pd.to_numeric(series, errors="coerce").dropna().to_numpy(float)
    median, q1, q3 = np.percentile(values, [50, 25, 75])
    return f"{median:.{decimals}f} [{q1:.{decimals}f}, {q3:.{decimals}f}]"

rows = []
for cluster in (0, 1, 2):
    group = indicators.loc[indicators["Cluster"] == cluster]
    evaluable = group.loc[group["indicator_evaluable"].astype(bool)]
    boundary_n = int(evaluable["t_star_boundary"].fillna(False).sum())
    rows.append(
        {
            "Cluster": f"C{cluster}",
            "Total n": len(group),
            "Evaluable n (%)": f"{len(evaluable)} ({100 * len(evaluable) / len(group):.1f}%)",
            "Boundary t* n (%)": f"{boundary_n} ({100 * boundary_n / len(evaluable):.1f}%)",
            "t*, months; median [Q1, Q3]": median_iqr(evaluable["operational_t_star"], 1),
            "max a(t); median [Q1, Q3]": median_iqr(evaluable["max_a_t"], 6),
            "mean a(t); median [Q1, Q3]": median_iqr(evaluable["mean_a_t"], 6),
        }
    )
panel_a = pd.DataFrame(rows)

# Rebuild the raw K-means labels used before canonical display relabeling.
verify_sha256(ANALYSIS_READY_FILE, EXPECTED_ANALYSIS_SHA256)
source = add_analysis_row_id(read_csv_robust(ANALYSIS_READY_FILE))
verify_rows(source, ANALYSIS_READY_N, "Analysis-ready dataset")
psa_columns = psa_columns_in_time_order(source)
psa = source[psa_columns].apply(pd.to_numeric, errors="coerce")
eligible = source.loc[((psa > 0) & np.isfinite(psa)).sum(axis=1) >= 2].copy()
verify_rows(eligible, TRAJECTORY_COHORT_N, "Eligible trajectory cohort")
with np.errstate(divide="ignore", invalid="ignore"):
    log_psa = np.log(eligible[psa_columns].apply(pd.to_numeric, errors="coerce").replace(0, np.nan))
change = log_psa.diff(axis=1)
smoothed = pd.DataFrame(
    np.vstack([
        gaussian_filter1d(row.to_numpy(float), sigma=REFERENCE_SIGMA, mode="nearest")
        for _, row in change.iterrows()
    ]),
    index=change.index,
    columns=change.columns,
)
features = pd.DataFrame(index=eligible.index)
features["max_change"] = smoothed.max(axis=1, skipna=True)
features["mean_change"] = smoothed.mean(axis=1, skipna=True)
features["first_positive_column_position"] = smoothed.apply(
    lambda row: int(np.argmax((row > 0).to_numpy(bool)))
    if np.any((row > 0).to_numpy(bool)) else -1,
    axis=1,
)
matrix = SimpleImputer(strategy="median").fit_transform(features)
raw_labels = KMeans(
    n_clusters=3, n_init=KMEANS_N_INIT, random_state=RANDOM_STATE
).fit_predict(matrix)
mapping_data = pd.DataFrame(
    {
        "Analysis_Row_ID": eligible["Analysis_Row_ID"].to_numpy(int),
        "Raw K-means label": raw_labels,
    }
).merge(clusters, on="Analysis_Row_ID", validate="one_to_one")
cross = pd.crosstab(mapping_data["Raw K-means label"], mapping_data["Cluster"])
mapping_rows = []
for raw_label in sorted(cross.index):
    display_label = int(cross.loc[raw_label].idxmax())
    if int(cross.loc[raw_label, display_label]) != int(cross.loc[raw_label].sum()):
        raise AssertionError("A raw cluster maps to more than one display cluster")
    mapping_rows.append(
        {
            "Raw K-means label": int(raw_label),
            "Reference display label": f"C{display_label}",
            "Patients, n": int(cross.loc[raw_label].sum()),
        }
    )
panel_b = pd.DataFrame(mapping_rows)

expected_a = [
    ["C0", 78, "78 (100.0%)", "3 (3.8%)", "12.0 [9.0, 18.0]", "0.194773 [0.092476, 0.404477]", "0.017347 [-0.003789, 0.049842]"],
    ["C1", 984, "962 (97.8%)", "353 (36.7%)", "18.0 [12.0, 24.0]", "0.252855 [0.105792, 0.426769]", "0.052227 [0.001385, 0.110827]"],
    ["C2", 372, "372 (100.0%)", "10 (2.7%)", "18.0 [15.0, 24.0]", "0.230981 [0.104545, 0.357568]", "0.018183 [-0.011698, 0.047239]"],
]
if panel_a.astype(object).values.tolist() != expected_a:
    raise AssertionError("Supplementary Table S3 panel A differs from the manuscript")
if panel_b.values.tolist() != [[0, "C1", 984], [1, "C2", 372], [2, "C0", 78]]:
    raise AssertionError(f"Supplementary Table S3 label mapping differs: {panel_b.values.tolist()}")

panel_a.to_csv(OUTDIR / "Supplementary_Table_S3A_Indicator_Evaluability.csv", index=False, encoding="utf-8-sig")
panel_b.to_csv(OUTDIR / "Supplementary_Table_S3B_Kmeans_Label_Mapping.csv", index=False, encoding="utf-8-sig")
markdown = (
    "# Supplementary Table S3. Figure 2 nonlinear-indicator evaluability and boundary assessment\n\n"
    + panel_a.to_markdown(index=False)
    + "\n\n## K-means label mapping\n\n"
    + panel_b.to_markdown(index=False)
    + "\n\nK-means numeric labels are arbitrary. The mapping changes labels only and does not alter patient membership. "
    "Percentages for boundary t* use indicator-evaluable patients as the denominator.\n"
)
(OUTDIR / "Supplementary_Table_S3.md").write_text(markdown, encoding="utf-8")

print("SUPPLEMENTARY TABLE S3 COMPLETION GATE PASSED")
print(panel_a.to_string(index=False))
print(panel_b.to_string(index=False))
